### PR TITLE
Add test of donor fields from subgraph without donor entity (#3160)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in sandbox and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: test/indexer/test_hca_indexer.py [hunk 1]
```diff
@@ -1279,6 +1279,91 @@ def test_multiple_samples(self):
...
+        hits = self._get_all_hits()
+        for hit in hits:
+            contents = hit['_source']['contents']
+            entity_type, aggregate = self._parse_index_name(hit)
+            if entity_type == 'projects':
+                if aggregate:
+                    # FIXME: https://github.com/databiosphere/azul/issues/3152
+                    #        The donor inner entity of a project aggregate should
+                    #        include None in each of the aggregated donor fields
+                    #        if one (or more) of the bundles has no donor entity.
```
Comment (achave11) [inline test/indexer/test_hca_indexer.py:1305]: You might want to add this as a second commit titled `Add FIXME (#3152)`.

File: test/indexer/test_hca_indexer.py [hunk 2]
```diff
@@ -1279,6 +1279,91 @@ def test_multiple_samples(self):
...
+                        #     'biological_sex': None,
+                        #     'genus_species': [None],
+                        #     'development_stage': None,
+                        #     'diseases': [None],
+                        #     'organism_age': None,
+                        #     'organism_age_value': None,
+                        #     'organism_age_unit': None
+                        # }
+                        # actual_donors = one(contents['donors'])
+                        # self.assertEqual(expected_donors, actual_donors)
```
Comment (achave11) [inline test/indexer/test_hca_indexer.py:1363]: Same.

**Review by achave11 — APPROVED** [review]
I think Hannes likes the FIXME commit (if it isn't already part of our process) because we can see references to where the changes are needed in the ticket. Not sure if this is an exception since the required FIXME changes are consolidated in a single test. Up to you, approved!

File: test/indexer/test_hca_indexer.py [hunk 3]
```diff
@@ -1276,6 +1276,91 @@ def test_multiple_samples(self):
...
+                            'organism_age_range': {
+                                'gte': 1955232000.0,
+                                'lte': 1955232000.0
+                            }
+                        }
+                        actual_donors = one(contents['donors'])
+                        self.assertEqual(expected_donors, actual_donors)
+                    elif sample_id == 'df23c109-59f0-46d3-bd09-660175b51bda':
+                        self.assertEqual(contents['donors'], [])
+                        # FIXME: Replace the assert above with the one below.
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1346]: See C-guide on disabling code.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Write the test so that it reproduces the issue. Push that separately as one commit. The CI build should fail. In a second commit, add the FIXME and tweak the tests with minimal conditionals to pass in presence of the bug.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Review comments should be addressed in a fixup commits. Find the commit that I reviewed previously in your reflog and recommit changes that address that review separately.

**Comment by hannes-ucsc** [comment]
Never mind.

File: test/indexer/test_hca_indexer.py [hunk 4]
```diff
@@ -1276,6 +1276,89 @@ def test_multiple_samples(self):
...
+                    else:
+                        for key, val in expected.items():
+                            if not isinstance(val, list):
+                                expected[key] = [val]
+                    expected.update(
+                        {
+                            'donor_count': 1,
+                            'donor_count_': 1
+                        }
+                    )
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1347]: It's confusing and inefficient that this is updating the expectation for every hit, in exactly the same way.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/indexer/test_hca_indexer.py [hunk 5]
```diff
@@ -1276,6 +1276,89 @@ def test_multiple_samples(self):
...
+        """
+        # Sample (Specimen): 70d2b85a, Donor: b111e5bf
+        bundle_fqid = self.bundle_fqid(uuid='1fd499c5-f397-4bff-9af0-eb42c37d5fbe',
+                                       version='2021-03-18T11:38:49.884Z')
+        self._index_canned_bundle(bundle_fqid)
+        # Sample (Organoid): df23c109, Donor: none
+        bundle_fqid = self.bundle_fqid(uuid='0722b70c-6778-423d-8fe9-869e2a515d35',
+                                       version='2021-03-18T11:38:49.863Z')
+        self._index_canned_bundle(bundle_fqid)
+        donor_b111e5bf = {
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1293]:
```suggestion
        donor = {
```

File: test/indexer/test_hca_indexer.py [hunk 6]
```diff
@@ -1276,6 +1276,89 @@ def test_multiple_samples(self):
...
+        donor_none = {
+            'document_id': None,
+            'biomaterial_id': None,
+            'biological_sex': None,
+            'genus_species': [None],
+            'development_stage': None,
+            'diseases': [None],
+            'organism_age': None,
+            'organism_age_value': None,
+            'organism_age_unit': None
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1317]: Why is organism_age_range missing from this literal?

With that exception, it seems like donor_none can be derived as well.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/indexer/test_hca_indexer.py (hunk 6 above)
Comment (danielsotirhos) [reply test/indexer/test_hca_indexer.py:1317]: `organism_age_range` is only added to donors inner entities if the donor has a `organism_age_in_seconds`. A donor that has no `organism_age` or `organism_age_unit` will have a `organism_age_in_seconds` of `None`, and thus no `organism_age_range`.

**Comment by danielsotirhos** [comment]
Note: Instead of a fixup commit, changes were amended to the original commits in order to preserve the failing commit & resolving FIXME commit.

File: test/indexer/test_hca_indexer.py [hunk 7]
```diff
@@ -1335,6 +1335,87 @@ def test_multiple_samples(self):
...
+        #        https://github.com/databiosphere/azul/issues/3152
+        if False:
+            aggregate_donor = {
+                key: val + [None] if isinstance(val, list) else [val, None]
+                for key, val in one(donor).items()
+            }
+            # The `organism_age_range` field will not have a `None`
+            # value since the field is only added to the donor
+            # inner entity if it has `organism_age_in_seconds` value.
+            aggregate_donor['organism_age_range'].pop()
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1395]: Brittle and obscure. It is not guaranteed that this pops the `None` entry and not immediately apparent.

File: test/indexer/test_hca_indexer.py [hunk 8]
```diff
@@ -1335,6 +1335,87 @@ def test_multiple_samples(self):
...
+            }
+        ]
+        # FIXME: The donor inner entity for a project without a donor
+        #        should have all the standard fields of a donor inner
+        #        entity with values of `None`.
+        #        https://github.com/databiosphere/azul/issues/3152
+        if False:
+            donor_none = [
+                {
+                    key: [None] if isinstance(val, list) else None
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1376]: Please stick to the established naming conventions.

File: test/indexer/test_hca_indexer.py [hunk 9]
```diff
@@ -1335,6 +1335,87 @@ def test_multiple_samples(self):
...
+        """
+        # Sample (Specimen): 70d2b85a, Donor: b111e5bf
+        bundle_fqid = self.bundle_fqid(uuid='1fd499c5-f397-4bff-9af0-eb42c37d5fbe',
+                                       version='2021-03-18T11:38:49.884Z')
+        self._index_canned_bundle(bundle_fqid)
+        # Sample (Organoid): df23c109, Donor: none
+        bundle_fqid = self.bundle_fqid(uuid='0722b70c-6778-423d-8fe9-869e2a515d35',
+                                       version='2021-03-18T11:38:49.863Z')
+        self._index_canned_bundle(bundle_fqid)
+        donor = [
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1352]: There is some plural vs singular confusion going on. If the value is a list a more appropriate name would be `donors`. Then again this is unpacked usign `one` twice but the list is only needed once.

File: test/indexer/test_hca_indexer.py [hunk 10]
```diff
@@ -1335,6 +1335,87 @@ def test_multiple_samples(self):
...
+            aggregate_donor = {
+                key: val + [None] if isinstance(val, list) else [val, None]
+                for key, val in one(donor).items()
+            }
+            # The `organism_age_range` field will not have a `None`
+            # value since the field is only added to the donor
+            # inner entity if it has `organism_age_in_seconds` value.
+            aggregate_donor['organism_age_range'].pop()
+        else:
+            aggregate_donor = {
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1397]: Not surgical enough,

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I'll push REVIEW commits.

This should make the FIXME commit a lot smaller.

I don't need to see fixups next time around.

File: test/indexer/test_hca_indexer.py [hunk 11]
```diff
@@ -1335,6 +1335,76 @@ def test_multiple_samples(self):
...
+                else:
+                    sample_id = one(contents['samples'])['document_id']
+                    if sample_id == '70d2b85a-8055-4027-a0d9-29452a49d668':
+                        self.assertEqual([donor], contents['donors'])
+                    elif sample_id == 'df23c109-59f0-46d3-bd09-660175b51bda':
+                        # FIXME: The donor inner entity for a project without a donor
+                        #        should have all the standard fields of a donor inner
+                        #        entity with values of `None`.
+                        #        https://github.com/databiosphere/azul/issues/3152
+                        self.assertEqual([donor_none] if False else [], contents['donors'])
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1404]:
```suggestion
                        self.assertEqual([] if True else [donor_none], contents['donors'])
```

To match the "disabler" for the other FIXME.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
One nit that's on me. No bump. No fixups needed next time.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**